src/CronExpression.php

```diff
@@ -4,155 +4,204 @@
-        if (strpos($time, '/') === 0)
-        {
-            $period = substr($time, 1);
-            return ($currentTime % $period) === 0;
-        }
-        // Handle ranges (1-5)
-        elseif (strpos($time, '-') !== false)
-        {
-            $items = [];
-            [$start, $end] = explode('-', $time);
-
-            for ($i = $start; $i <= $end; $i++)
-            {
-                $items[] = $i;
-            }
-        }
-        // Handle multiple days
-        else
-        {
-            $items = explode(',', $time);
-        }
-
-        return in_array($currentTime, $items);
-    }
+	/**
+	 * The timezone this should be considered under.
+	 *
+	 * @var string
+	 */
+	protected $timezone;
+
+	/**
+	 * The current date/time. Used for testing.
+	 *
+	 * @var \DateTime
+	 */
+	protected $testTime;
+
+	/**
+	 * Allows us to set global timezone for all tasks
+	 * on construct
+	 *
+	 * @param string $timezone The global timezone for all tasks
+	 *
+	 * @return void
+	 */
+	public function __construct(string $timezone = null)
+	{
+		if ($timezone === null)
+		{
+			$timezone = app_timezone();
+		}
+
+		$this->timezone($timezone);
+	}
+
+	/**
+	 * Checks whether cron expression should be run. Allows
+	 * for custom timezone to be used for specific task
+	 *
+	 * @param string $expression The Cron Expression to be evaluated
+	 * @param string $timezone   The timezone for this specific task
+	 *
+	 * @return boolean
+	 */
+	public function shouldRun(string $expression, string $timezone = null): bool
+	{
+		// Set our current time
+		if (! $this->testTime instanceof Time)
+		{
+			$this->testTime = Time::now($this->timezone);
+		}
+
+		if (! is_null($timezone))
+		{
+			// Convert time object to selected timezone
+			$runtimeTimezone = new \DateTimeZone($timezone);
+		}
+		else
+		{
+			$runtimeTimezone = $this->timezone;
+		}
+
+		$this->testTime = $this->testTime->setTimezone($runtimeTimezone);
+
+		// Break the expression into separate parts
+		[
+			$min,
+			$hour,
+			$monthDay,
+			$month,
+			$weekDay,
+		] = explode(' ', $expression);
+
+		return $this->checkMinute($min)
+			&& $this->checkHour($hour)
+			&& $this->checkMonthDay($monthDay)
+			&& $this->checkMonth($month)
+			&& $this->checkWeekDay($weekDay);
+	}
+
+	/**
+	 * Returns a Time instance representing the next
+	 * date/time this expression would be ran.
+	 *
+	 * @param string $expression
+	 *
+	 * @return Time
+	 */
+	public function nextRun(string $expression): Time
+	{
+	}
+
+	/**
+	 * Returns a Time instance representing the last
+	 * date/time this expression would have ran.
+	 *
+	 * @param string $expression
+	 *
+	 * @return Time
+	 */
+	public function lastRun(string $expression): Time
+	{
+	}
+
+	/**
+	 * Sets the timezone that dates should be examined under.
+	 *
+	 * @param string $timezone
+	 *
+	 * @return $this
+	 * @throws \Exception Thrown if $timezone is not a valid timezone string.
+	 */
+	public function timezone(string $timezone)
```

Comment:
please rename to `setTimezone` for clarity. Also - I think the logic you added into `shouldRun()` dealing with timezones could be placed here, allowing `shouldRun` to focus on it's one task, and to remove extra arguments.

Reply:
Do you mean lines 56-66? If we move that to setTimezone we would lose the ability to add custom timezones at run time. Doing something like
```
$cron = new CronExpression();
$cron ->setTimezone('America/Chicago');
$cron->shouldRun('* 10 * * *'); // Checks Chicago Time
$cron->shouldRun('* 10 * * *','Europe/Athens'); // Checks Athens
$cron->shouldRun('* 10 * * *'); // Checks Chicago again
```

would need setTimezone() to be called anytime we needed to switch timezones. If you don't think such a functionality is not needed I can move all that logic to setTimezone, sure.

Reply:
It was late last night when I looked at this, and my coffee hasn't fully kicked in yet this morning :) but let me see if I can understand. The default timezone should be `Config\App::$appTimezone`, right? I guess we're not taking that into account, currently. When I was looking at it lastnight, I was thinking it would be used like? 

```
$cron = new CronExpression();
$cron->setTimezone('Europe/Athens')->shouldRun(...);
```

Basically, I was trying to minimize argument usage since I feel that's clearer, but as I think through it I think what you've done here is fine. It allows us to pass in the base timezone as a dependency through `setTimezone` which makes testing easier anyway, and the additional argument makes sense. Besides, this class is an internal file, anyway. 

Long story short: ignore me lol

Reply:
No worries. Minimizing arguments is not an invalid argument.
The default timezone should be Config\App::$appTimezone, you're right. I used app_timezone() instead, I'll fix it.

Regarding the shouldRun() function, I think the main thing we need to figure out is how the CronExpression class is going to be used and how it will be connected to the Tasks.

Is it going to be simply a "helper" class?
For example if Tasks is something like
```php
class Task{

    public function __construct($job,$expression,$timezone){
        //
        // Pretend we handle args here
        //

        $this->cron = new CronExpression();
        $this->cron->setTimezone($this->timezone);
    }

    public function run(){
        if($this->cron->shouldRun($this->expression)){
            // Do Job
        }
    }

    // Whatever other logic is needed
}


```

then setting a timezone at shouldRun should not be needed.

If instead we want to do something more complicated, like assign each task multiple timezones, then it might be helpful to be able to assign the timezone on shouldRun().

That being said, I think it's a really extreme edge case that a Task would need two different timezones. I can't honestly think of any examples where that would be needed. I'll move everything to setTimezone and fix the tests.

Reply:
I had forgotten you were using app_timezone() already. No need to change. 

The intent is that this class is only used internally and not exposed publicly. I cannot think of a reason we would ever need multiple timezones for a single task. Not something I want to support either :) On the weird case they actually come up with a use for that, they can break it up into multiple jobs lol. 

I'll try to get some basic docs written up very soon so that we know how the pieces should fit together.